src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -605,9 +605,11 @@ private void UnwrapCollectionExpressionIfNullable(ref BoundExpression collection
         /// <param name="builder">Builder to fill in (partially, all but conversions).</param>
         /// <param name="collectionExpr">The expression over which to iterate.</param>
         /// <param name="diagnostics">Populated with binding diagnostics.</param>
+        /// <param name="isAsyncOverride">In error cases, we'll retry with a different assumption of whether the foreach is async or not.</param>
         /// <returns>Partially populated (all but conversions) or null if there was an error.</returns>
-        private bool GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder builder, BoundExpression collectionExpr, DiagnosticBag diagnostics)
+        private bool GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder builder, BoundExpression collectionExpr, bool? isAsyncOverride, DiagnosticBag diagnostics)
```

Comment:
Rather than passing a nullable parameter, could you make a helper method instead that tries twice for the times when it's appropriate?